opm/simulators/flow/BlackoilModelEbos.hpp

```diff
@@ -245,6 +245,9 @@ namespace Opm {
             convergence_reports_.reserve(300); // Often insufficient, but avoids frequent moves.
             // TODO: remember to fix!
             if (param_.nonlinear_solver_ == "nldd") {
+                if (!param_.matrix_add_well_contributions_) {
+                    OPM_THROW(std::runtime_error, "The --nonlinear-solver=nldd option can only be used with --matrix-add-well-contributions=true");
+                }
```

Comment:
Does this shut down nicely in parallel runs or do we get one of those ugly messages from the MPI system?

Reply:
I think every rank throws here since the parameters are common and this is in the constructor of the main model object. But I still get a somewhat ugly message:
```
Error: [/Users/atgeirr/opm/src/opm-simulators/opm/simulators/flow/BlackoilModelEbos.hpp:249] The --nonlinear-solver=nldd option can only be used with --matrix-add-well-contributions=true
Simulation aborted as program threw an unexpected exception: [/Users/atgeirr/opm/src/opm-simulators/opm/simulators/flow/BlackoilModelEbos.hpp:249] The --nonlinear-solver=nldd option can only be used with --matrix-add-well-contributions=true
Restart file written for report step   0/247, date = 06-Nov-1997 00:00:00
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[8198,1],1]
  Exit code:    1
--------------------------------------------------------------------------
```
Not sure that is the fault of this code, I think I see this whenever I get an exception-triggered exit. Maybe we do not call some MPI cleanup properly?

Reply:
> I think I see this whenever I get an exception-triggered exit

Yes we do get this kind of message in that case.

> Maybe we do not call some MPI cleanup properly?

I believe that's the crux of the problem, yes.  We don't quite properly capture the failure condition in a way that allows [`MPI_Finalize()`](https://github.com/OPM/opm-simulators/blob/68c1b45f13093026755ec50393c70280655b24f3/opm/simulators/flow/Main.cpp#L110) to properly shut down the system.  Fixing that, though, is probably a somewhat sizeable undertaking so not something we should do right now.